### ISSUE TITLE
Wrong SQL generated on query on deeply nested association

### ISSUE THREAD

**Issue by HolySamosa** [sequelize/sequelize#5353]
I'm attempting to execute a query which is filtering on a 3-deep nested association.  It seems that Sequelize is generating the wrong SQL for the WHERE clause on the INNER JOIN.  Sorry, I didn't remove the likely superfluous members, just in case they are affecting anything.

Here's the query:

```
        return models.ActiveShift.find({
          order: ['priority', 'lastActivity'],
          include: [{model: models.Messageable, as: 'operator', include: {model: models.Case, as: 'operatorCases', where: {id: 42 /*the case id I'm searching for*/}}}]
        })
```

With these models:

```
  var ActiveShift = sequelize.define("ActiveShift", {

    clockedIn: {type: DataTypes.DATE, defaultValue: DataTypes.NOW, allowNull: false},
    lastActivity: {type: DataTypes.DATE, defaultValue: DataTypes.NOW, allowNull: false},
    priority: {type: DataTypes.INTEGER, defaultValue: 1, allowNull: false}
  }, {
    timestamps: false,
    classMethods: {
      associate: function (models) {
        ActiveShift.belongsToMany(models.Case, { as: "activeCases", through: "ActiveShiftCase"});
        ActiveShift.belongsTo(models.Messageable, { as: "operator", foreignKey: {unique: true}, onDelete: 'CASCADE'});
      },
```

```
  var Messageable = sequelize.define("Messageable", {
    handle: {type: DataTypes.STRING, unique: true},
    mobile: {
      type: DataTypes.STRING,
      allowNull:false,
      validate: { is:{ args:[/^\+1\d{10}$/], msg: 'Must be a US phone number in E.164 format.'}}
    },
    type: {type: DataTypes.ENUM, values: ['operator', 'texter'], allowNull: false}
  }, {
    timestamps: false,
    classMethods: {
      associate: function (models) {
        // only applies to messagable type = operator
        Messageable.belongsToMany(models.Case, {as: "operatorCases", through: "CaseOperator"});
        // only applies to messageable type = texter
        Messageable.hasOne(models.Case, {as: 'texterCase', foreignKey : 'texterId'});
      },
```

```
 var Case = sequelize.define("Case", {

    }, {
      classMethods: {
        associate: function (models) {

          Case.belongsToMany(models.Messageable, {as: "operators", through: "CaseOperator"});
          Case.belongsTo(models.Messageable, { as: "texter"});
        },
```

Here's the resulting SQL:

```
SELECT
   `ActiveShift`.*,
   `operator.operatorCases`.`id` AS `operator.operatorCases.id`,
   `operator.operatorCases`.`createdAt` AS `operator.operatorCases.createdAt`,
   `operator.operatorCases`.`updatedAt` AS `operator.operatorCases.updatedAt`,
   `operator.operatorCases`.`texterId` AS `operator.operatorCases.texterId`,
   `operator.operatorCases.CaseOperator`.`createdAt` AS `operator.operatorCases.CaseOperator.createdAt`,
   `operator.operatorCases.CaseOperator`.`updatedAt` AS `operator.operatorCases.CaseOperator.updatedAt`,
   `operator.operatorCases.CaseOperator`.`MessageableId` AS `operator.operatorCases.CaseOperator.MessageableId`,
   `operator.operatorCases.CaseOperator`.`CaseId` AS `operator.operatorCases.CaseOperator.CaseId` 
FROM
   (SELECT
      `ActiveShift`.`id`,
      `ActiveShift`.`clockedIn`,
      `ActiveShift`.`lastActivity`,
      `ActiveShift`.`priority`,
      `ActiveShift`.`operatorId`,
      `operator`.`id` AS `operator.id`,
      `operator`.`handle` AS `operator.handle`,
      `operator`.`mobile` AS `operator.mobile`,
      `operator`.`type` AS `operator.type` 
   FROM
      `ActiveShifts` AS `ActiveShift` 
   LEFT OUTER JOIN
      `Messageables` AS `operator` 
         ON `ActiveShift`.`operatorId` = `operator`.`id` 
   WHERE
      (
         SELECT
            `Messageable`.`id` 
         FROM
            `Messageables` AS `Messageable` 
         INNER JOIN
            `CaseOperator` AS `operatorCases.CaseOperator` 
               ON `Messageable`.`id` = `operatorCases.CaseOperator`.`MessageableId` 
         INNER JOIN
            `Cases` AS `operatorCases` 
               ON `operatorCases`.`id` = `operatorCases.CaseOperator`.`CaseId` 
               AND `operatorCases`.`id` = 1 
         WHERE
            `ActiveShift`.`id` = `Messageable`.`operatorId` LIMIT 1 
      ) IS NOT NULL LIMIT 1
   ) AS `ActiveShift` 
INNER JOIN
   `CaseOperator` AS `operator.operatorCases.CaseOperator` 
      ON `operator.id` = `operator.operatorCases.CaseOperator`.`MessageableId` 
INNER JOIN
   `Cases` AS `operator.operatorCases` 
      ON `operator.operatorCases`.`id` = `operator.operatorCases.CaseOperator`.`CaseId` 
      AND `operator.operatorCases`.`id` = 1 
ORDER BY
   `priority`,
   `lastActivity`;
```

The query results in this error: `SQLITE_ERROR: no such column: Messageable.operatorId`.

It seems to me that the WHERE clause should be:

```
         WHERE
            `ActiveShift`.`operatorId` = `Messageable`.`id` LIMIT
```

Basically, the column names are switched between the two tables.  ActiveShift does have an operatorId, while Messageable does not.  If I execute the raw SQL with the described modification of the where clause, I get the expected result. 

If this is a bug, what work-around would be recommended? I briefly looked at using a raw query for the entire thing, however it didn't seem that I would be able to get eager loading of the associations, which I need (only 1-deep, activeShift.operator, should that make a difference).

**Comment by HolySamosa** [comment]
Ok, looks like I was mistaken about raw queries-- seems I should be able to eager load the nested models.  I tried working around with this query:
```
        sequelize.query(
            // rawQueryString created based on fixed raw query above
            rawQueryString, {
            replacements: {caseId: searchingForThisId},
            type: Sequelize.QueryTypes.SELECT,
            model: {
              model: models.ActiveShift,
              include: [{model: models.Messageable, as: 'operator'}]
            }
          })
```

but get this error:
```
TypeError: this.model.bulkBuild is not a function
    at AbstractQuery.handleSelectQuery (c:\Projects Source\SafeLine\node_modules\sequelize\lib\dialects\abstract\query.js:682:25)
    at Statement.Query.run.self.database.serialize.afterExecute (c:\Projects Source\SafeLine\node_modules\sequelize\lib\dialects\sqlite\query.js:178:33)
--> in Database#all('SELECT `ActiveShift`.*, `operator.operatorCases`.`id` ...
```

Suggestions?

**Comment by HolySamosa** [comment]
I discovered that I can avoid this issue all together by leveraging some earlier query results to simplify my query.  However it does seem that there is a bug in Sequelize resulting in an erroneous WHERE clause.

**Comment by mickhansen** [comment]
Looks like a bug, i'd be very grateful if you could supply a pull request replicating this error in as little code as possible.

**Comment by HolySamosa** [comment]
@mickhansen Will do.  Thanks for the great work!

**Comment by xxxtonixxx** [comment]
I have the same problem, any news?
